apollo-federation/src/sources/connect/json_selection/selection_trie.rs

```diff
@@ -0,0 +1,390 @@
+                (key, sub).hash(&mut hasher);
+                acc ^ hasher.finish()
+            })
+            .hash(state);
+    }
+}
+
+impl SelectionTrie {
+    pub(crate) fn new() -> Self {
+        Self {
+            used: false,
+            selections: IndexMap::default(),
+            key_ranges: IndexMap::default(),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn new_used() -> Self {
+        Self {
+            used: true,
+            selections: IndexMap::default(),
+            key_ranges: IndexMap::default(),
+        }
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.selections.is_empty()
+    }
+
+    pub(crate) fn keys(&self) -> impl Iterator<Item = &String> {
+        self.selections.keys()
+    }
+
+    pub(crate) fn get(&self, key: impl Into<String>) -> Option<&SelectionTrie> {
+        self.selections.get(&key.into()).map(|sub| sub.as_ref())
+    }
+
+    pub(crate) fn iter(&self) -> impl Iterator<Item = (&str, &SelectionTrie)> {
+        self.selections
+            .iter()
+            .map(|(key, sub)| (key.as_str(), sub.as_ref()))
+    }
+
+    pub(crate) fn key_ranges(&self, key: &str) -> impl Iterator<Item = Range<usize>> {
+        self.key_ranges
+            .get(key)
+            .into_iter()
+            .flat_map(|ranges| ranges.iter())
+            .cloned()
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn has_str_path<'a>(&self, path: impl IntoIterator<Item = &'a str>) -> bool {
+        let mut current = self;
+        for key in path {
+            if let Some(sub) = current.get(key) {
+                current = sub;
+            } else {
+                return false;
+            }
+        }
+        current.is_used()
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn add_str_path<'a>(
+        &mut self,
+        path: impl IntoIterator<Item = &'a str>,
+    ) -> &mut Self {
+        path.into_iter()
+            .fold(self, |trie, key| trie.add_str(key))
+            .set_used()
+    }
+
+    pub(super) fn add_path_list(&mut self, path_list: &PathList) -> &mut Self {
+        match path_list {
+            PathList::Key(key, tail) => self.add_key(key).add_path_list(tail.as_ref()),
+            PathList::Selection(sub) => self.add_subselection(sub),
+            // If we get to the end of the PathList, mark the path used.
+            PathList::Empty => self.set_used(),
+            // TODO Support PathList::Method and inputs used within method
+            // arguments. For now, assume we use the whole path up to the
+            // unhandled PathList element.
+            _ => self.set_used(),
+        }
+    }
+
+    pub(crate) fn add_subselection(&mut self, sub: &SubSelection) -> &mut Self {
+        for selection in sub.selections_iter() {
+            match selection {
+                NamedSelection::Field(_, key, nested_selection) => {
+                    let result = self.add_key(key);
```

Comment:
where is this used after if else? feels like dead code

Reply:
Definitely not dead code: `self.add_key(key)` returns a mutable reference to the child `SelectionTrie` that was either found or created, so the `result.add_subselection(nested)` and `result.set_leaf()` calls are updating it.